src/protocols/output_management.rs

```diff
@@ -0,0 +1,791 @@
+            display,
+            client_data.manager.version(),
+            output.name.clone(),
+        )
+        .unwrap();
+    client_data.manager.head(&new_head);
+    new_head.name(output.name.clone());
+    new_head.description(output.name.clone());
+    if let Some((width, height)) = output.physical_size {
+        if let (Ok(a), Ok(b)) = (width.try_into(), height.try_into()) {
+            new_head.physical_size(a, b);
+        }
+    }
+    let mut index = 0;
+    let mut new_modes = Vec::with_capacity(output.modes.len());
+    for mode in &output.modes {
+        let new_mode = client
+            .create_resource::<ZwlrOutputModeV1, _, D>(display, new_head.version(), ())
+            .unwrap();
+        new_head.mode(&new_mode);
+        if let (Ok(width), Ok(height)) = (mode.width.try_into(), mode.height.try_into()) {
+            new_mode.size(width, height);
+        }
+        if mode.is_preferred {
+            new_mode.preferred();
+        }
+        if let Ok(refresh_rate) = mode.refresh_rate.try_into() {
+            new_mode.refresh(refresh_rate);
+        }
+        if Some(index) == output.current_mode {
+            new_head.current_mode(&new_mode);
+        }
+        new_modes.push(new_mode);
+        index += 1;
+    }
+    if let Some(logical) = output.logical {
+        new_head.position(logical.x, logical.y);
+        new_head.transform(ipc_transform_to_smithay(logical.transform).into());
+        new_head.scale(logical.scale);
+    }
+    new_head.enabled(output.current_mode.is_some() as i32);
+    new_head.make(output.make.clone());
+    new_head.model(output.model.clone());
+
+    new_head.adaptive_sync(match output.vrr_enabled {
+        true => AdaptiveSyncState::Enabled,
+        false => AdaptiveSyncState::Disabled,
+    });
+    // new_head.serial_number(output.serial);
+    client_data
+        .heads
+        .insert(output.name.clone(), (new_head, new_modes));
+}
+
+pub fn notify_changes(state: &mut OutputManagementManagerState, new_out: IpcOutputMap) {
+    let mut changed = false; /* most likely to endup true */
+    for (name, new) in new_out.iter() {
+        if let Some(old) = state.current_out.get(name) {
+            if old.vrr_enabled != new.vrr_enabled {
+                changed = true;
+                for (_, client) in &state.clients {
+                    if let Some((head, _)) = client.heads.get(name) {
+                        head.adaptive_sync(match new.vrr_enabled {
+                            true => AdaptiveSyncState::Enabled,
+                            false => AdaptiveSyncState::Disabled,
+                        });
+                    }
+                }
+            }
+            match (old.current_mode, new.current_mode) {
+                (Some(old_index), Some(new_index)) => {
+                    if old.modes.len() <= old_index
+                        || old.modes.len() <= new_index
+                        || new.modes.len() <= new_index
```

Comment:
Could you just check `old.modes == new.modes` and error!() out otherwise? I am not sure modes can change over the lifetime of an output.